### ISSUE TITLE
MC format sstable writer accumulates all promoted index entries

### ISSUE THREAD

**Issue by tgrabiec** [scylladb/scylla#4217]
In sstables/mc/writer.cc:
```
        // Unfortunately we cannot output the promoted index directly to the
        // index file because it needs to be prepended by its size.
        seastar::circular_buffer<pi_block> promoted_index;
```

I saw:

```
WARN  2019-02-12 19:58:48,930 [shard 0] seastar_memory - oversized allocation: 1310720 bytes, please report:
 (inlined by) sstables::mc::writer::add_pi_block() at /mnt/src/scylla/sstables/mc/writer.cc:752
sstables::mc::writer::consume(clustering_row&&) at /mnt/src/scylla/sstables/mc/writer.cc:763
 (inlined by) ?? at /mnt/src/scylla/sstables/mc/writer.cc:664
 (inlined by) sstables::mc::writer::consume(clustering_row&&) at /mnt/src/scylla/sstables/mc/writer.cc:1212
void flat_mutation_reader::impl::consume_pausable_in_thread<std::reference_wrapper<flat_mutation_reader::impl::consumer_adapter<stable_flattened_mutations_consumer<compact_for_compaction<sstables::compacting_sstable_writer> > > >, std::function<bool (dht::decorated_key const&)> >(std::reference_wrapper<flat_mutation_reader::impl::consumer_adapter<stable_flattened_mutations_consumer<compact_for_compaction<sstables::compacting_sstable_writer> > > >, std::function<bool (dht::decorated_key const&)>, std::chrono::time_point<seastar::lowres_clock, std::chrono::duration<long, std::ratio<1l, 1000l> > >) at /mnt/src/scylla/sstables/compaction.cc:145

```

For large partitions can cause OOM and/or compaction failure.

**Comment by avikivity** [comment]
>        // Unfortunately we cannot output the promoted index directly to the
>        // index file because it needs to be prepended by its size.

This is terrible. Even if we replace the circular_buffer with a chunked_vector, it can still consume a lot of memory.

I think we should limit the amount of memory consumed by the promoted index and do a RMW at the offset if the limit is exceeded. This is hard because the file_output_stream's write-behind cache is not coherent.

**Comment by avikivity** [comment]
For ka I proposed increasing desired_block_size so that growth is slowed (it's still not bounded), but it doesn't make sense here, because it will reduce the binary search's effectiveness. For ka we only had linear search.

**Comment by avikivity** [comment]
Anyway, chunked_vector can provide relief for most cases.

**Comment by tgrabiec** [comment]
Another problem with going back and serializing the size after the fact is that those sizes are vints. However, I think we can always force that vint to a fixed-size full-precision uint64. If a partition is large enough to require an index, sparing a few bytes on per partition shouldn't significantly affect the on-disk size.

We could have some support for this in `output_stream` and `file_data_sink_impl`, which would allow writing a placeholder, returning an object which can be used to fill it. A buffer with active placeholders would not be written back by write-behind, only when all placeholders are filled-in. `flush()` with an active placeholder would throw.

We could use the same mechanism for writing row flags, to avoid double-pass where we first determine the flags based on row contents and then do the same logic to write the content.

**Comment by tzach** [comment]
@avikivity @tgrabiec can we close this issue?

**Comment by tgrabiec** [comment]
@tzach No, we still accumulate all entries in memory, which may lead to OOM. Some ideas for solutions were posted here.

**Comment by bhalevy** [comment]
@avikivity do we want this in 5.0 / 2022.1?

**Comment by avikivity** [comment]
Reopening. The fix reduced the magnitude of the problem but we still keep the promoted index in memory.

Probably not fixable without major effort, and not worth fixing, but still, it's not fixed.

**Comment by avikivity** [comment]
As to backports, this isn't a regression.

**Comment by bhalevy** [comment]
@avikivity the issue wasn't reopened, so reopening now.

**Comment by avikivity** [comment]
> @avikivity the issue wasn't reopened, so reopening now.

Oops :(

**Comment by mykaul** [comment]
ping @avikivity (for backport ? unsure if it's not too big?)

**Comment by mykaul** [comment]
> ping @avikivity (for backport ? unsure if it's not too big?)

ping @avikivity - unsure if it's still relevant even?

**Comment by avikivity** [comment]
It's still relevant.

It was somewhat mitigated by 19316dee94bf9cef6bcf2ae5f82078da9d78434a though. So we have less risk of OOM at the cost of a worse index.

**Comment by raphaelsc** [comment]
> It's still relevant.
> 
> It was somewhat mitigated by [19316de](https://github.com/scylladb/scylladb/commit/19316dee94bf9cef6bcf2ae5f82078da9d78434a) though. So we have less risk of OOM at the cost of a worse index.

Note doubling the promoted index block size might reduce memory pressure at the time of writing, but might cause it later at the time of reading when doing the binary search over the blocks. Probably our best solution is about splitting sstables at clustering boundary, so the memory requirement becomes a max of 1G/64*sizeof(promoted_index_block).

**Comment by avikivity** [comment]
Yes I forgot about that. It's a good idea.

Even accumulating a 1GB sstable's promoted index can be bad in extreme cases, but usually splitting a partition will alleviate much of the problem.
